File: batch-cancelled-partial-results/python/batch_cancellation_audit.py
"""Find billed, salvageable output left behind by cancelled batches.

Read only, on both providers. Two GET endpoints, /v1/batches on OpenAI and
/v1/messages/batches on Anthropic. This script never cancels a batch, never
submits one, and never downloads a result file.

Cancel is a stop, not a rollback. Requests that reached the model before the
cancel landed are finished and are in the output; re-running the whole batch
pays for them twice. Anthropic documents that canceled and expired requests are
not billed. OpenAI documents the partial output but not the billing split, so
the completed count is reported as a floor rather than as a total.

Retention of that partial output, and the join against your own ingest ledger,
belong to the unclaimed-output note. This one stops at the arithmetic.
"""
import argparse
import calendar
import datetime
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("batch_cancellation_audit")

OPENAI_BATCHES_URL = "https://api.openai.com/v1/batches"
ANTHROPIC_BATCHES_URL = "https://api.anthropic.com/v1/messages/batches"

# OpenAI documents up to ten minutes in "cancelling" before a batch reaches
# "cancelled". Fifteen is a generous read of ten. Anthropic publishes no bound
# for "canceling" at all, so on that side this is a heuristic and the output
# says so rather than dressing a guess up as a rule.
STUCK_SECONDS = 15 * 60

OPENAI_CANCEL_STATES = ("cancelling", "cancelled")

FINDINGS = ("cancel-stuck", "cancel-partial-unclaimed")


def parse_time(value):
    """Epoch seconds from a unix number or an RFC 3339 string. Pure.

    OpenAI stamps integers, Anthropic stamps RFC 3339. Everything downstream
    wants one kind of number, so the difference is absorbed once, here. A
    string with no offset is read as UTC rather than as whatever the machine
    running the audit happens to be set to.
    """
    if value is None or value == "" or isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return int(value)
    text = str(value).strip()
    if text[-1:] in ("Z", "z"):
        text = text[:-1] + "+00:00"
    try:
        stamp = datetime.datetime.fromisoformat(text)
    except (ValueError, TypeError):
        return None
    if stamp.tzinfo is None:
        return calendar.timegm(stamp.timetuple())
    return int(stamp.timestamp())


def openai_cancel_rows(batches):
    """Normalised rows for OpenAI batches under cancellation. Pure."""
    out = []
    for b in batches or []:
        status = (b or {}).get("status")
        if status not in OPENAI_CANCEL_STATES:
            continue
        counts = b.get("request_counts") or {}
        total = int(counts.get("total") or 0)
        done = int(counts.get("completed") or 0)
        failed = int(counts.get("failed") or 0)
        out.append({
            "provider": "openai",
            "id": str(b.get("id")),
            "status": status,
            "in_flight": status == "cancelling",
            "done": done,
            "stopped": max(0, total - done - failed),
            "total": total,
            "artifact": b.get("output_file_id"),
            "cancel_started": parse_time(b.get("cancelling_at")),
            "billing_known": False,
        })
    return sorted(out, key=lambda r: r["id"])


def anthropic_cancel_rows(batches):
    """Normalised rows for Claude batches under cancellation. Pure.

    cancel_initiated_at is set only when cancellation was initiated, and stays
    set once the batch has ended, so it is the whole filter.
    """
    out = []
    for b in batches or []:
        started = (b or {}).get("cancel_initiated_at")
        if not started:
            continue
        counts = b.get("request_counts") or {}
        done = int(counts.get("succeeded") or 0)
        stopped = int(counts.get("canceled") or 0)
        status = str(b.get("processing_status") or "")
        out.append({
            "provider": "anthropic",
            "id": str(b.get("id")),
            "status": status or "unknown",
            "in_flight": status == "canceling",
            "done": done,
            "stopped": stopped,
            "total": done + stopped + int(counts.get("errored") or 0)
                     + int(counts.get("expired") or 0)
                     + int(counts.get("processing") or 0),
            "artifact": b.get("results_url"),
            "cancel_started": parse_time(started),
            "billing_known": True,
        })
    return sorted(out, key=lambda r: r["id"])


def salvage_rows(rows):
    """Rows holding finished work a re-run would pay for again. Pure."""
    return [r for r in rows or [] if int(r.get("done") or 0) > 0]


def stuck_rows(rows, now, seconds=STUCK_SECONDS):
    """Rows still mid cancel past the threshold. Pure. now is an argument."""
    out = []
    for r in rows or []:
        if not r.get("in_flight"):
            continue
        started = r.get("cancel_started")
        if started is None or now - started > seconds:
            out.append(r)
    return out


def salvaged_total(rows):
    """Finished rows across everything cancelled. Pure."""
    return sum(int(r.get("done") or 0) for r in salvage_rows(rows))


def verdict(rows, stuck, salvage):
    """Grade the run. Pure. Returns (state, detail)."""
    rows = list(rows or [])
    stuck = list(stuck or [])
    salvage = list(salvage or [])
    if not rows:
        return ("no-cancels",
                "no batch on the providers checked has had a cancellation "
                "initiated")
    if stuck:
        detail = ("%d batch(es) have been mid cancel longer than the documented "
                  "window" % len(stuck))
        if salvage:
            detail += (", and %d cancelled batch(es) hold %d finished rows "
                       "nothing has collected"
                       % (len(salvage), salvaged_total(salvage)))
        return ("cancel-stuck", detail)
    if salvage:
        return ("cancel-partial-unclaimed",
                "%d cancelled batch(es) hold %d finished rows a re-run would "
                "pay for again" % (len(salvage), salvaged_total(salvage)))
    return ("cancel-clean",
            "%d cancellation(s) found, none of which had completed a single "
            "request, so there is nothing to salvage and nothing to double pay"
            % len(rows))


def repair_lines(state, rows):
    """The repair for one verdict. Pure. Printed, never performed."""
    rows = list(rows or [])
    if state == "no-cancels":
        return []
    if state == "cancel-clean":
        return ["nothing to collect. Keep cancelling early: a batch stopped "
                "before its first request completed costs nothing."]
    lines = []
    if state == "cancel-stuck":
        lines.append("a batch still in cancelling or canceling has not stopped. "
                     "Poll it to a terminal state before you submit a "
                     "replacement, or the two will run the same rows at once.")
    if any(int(r.get("done") or 0) > 0 for r in rows):
        lines.append("download the partial output, drop those custom_ids from "
                     "the input file, and submit only the remainder. Results "
                     "are not returned in request order, so custom_id is the "
                     "only join key that works.")
    if any(r.get("provider") == "anthropic" and int(r.get("done") or 0) > 0
           for r in rows):
        lines.append("on Anthropic, canceled and expired requests are not "
                     "billed, so the succeeded count is the whole cost of the "
                     "cancelled batch.")
    if any(r.get("provider") == "openai" and int(r.get("done") or 0) > 0
           for r in rows):
        lines.append("on OpenAI the billing split for a cancelled batch is not "
                     "documented, so treat the completed count as a floor and "
                     "confirm the day against the cost report.")
    return lines


def get_json(url, headers, params=None, timeout=30):
    """One GET. Returns (payload, error). Read only, always."""
    try:
        r = requests.get(url, headers=headers, params=params or {}, timeout=timeout)
    except requests.RequestException as exc:
        return (None, "request failed: %s" % exc)
    if r.status_code != 200:
        return (None, "HTTP %d %s" % (r.status_code, (r.text or "")[:160]))
    try:
        return (r.json(), None)
    except ValueError:
        return (None, "response was not JSON")


def page_openai(key, max_pages):
    """/v1/batches, following the after cursor. GETs only."""
    rows, after = [], None
    headers = {"Authorization": "Bearer %s" % key}
    for _ in range(max(1, max_pages)):
        params = {"limit": 100}
        if after:
            params["after"] = after
        payload, err = get_json(OPENAI_BATCHES_URL, headers, params)
        if err:
            return (rows, err)
        data = payload.get("data") or []
        rows.extend(data)
        if not payload.get("has_more") or not data:
            break
        after = data[-1].get("id")
    return (rows, None)


def page_anthropic(key, max_pages):
    """/v1/messages/batches, following after_id. GETs only."""
    rows, after = [], None
    headers = {"x-api-key": key, "anthropic-version": "2023-06-01"}
    for _ in range(max(1, max_pages)):
        params = {"limit": 1000}
        if after:
            params["after_id"] = after
        payload, err = get_json(ANTHROPIC_BATCHES_URL, headers, params)
        if err:
            return (rows, err)
        data = payload.get("data") or []
        rows.extend(data)
        if not payload.get("has_more") or not data:
            break
        after = payload.get("last_id") or data[-1].get("id")
    return (rows, None)


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--stuck-minutes", type=int, default=15,
                    help="age past which a batch still mid cancel is reported")
    ap.add_argument("--max-pages", type=int, default=20)
    args = ap.parse_args()

    openai_key = os.environ.get("OPENAI_API_KEY")
    anthropic_key = os.environ.get("ANTHROPIC_API_KEY")
    if not openai_key and not anthropic_key:
        log.error("set OPENAI_API_KEY (project key, Read Only) or "
                  "ANTHROPIC_API_KEY (workspace key), or both")
        return 2

    now = int(time.time())
    rows = []
    checked = []
    if openai_key:
        checked.append("openai")
        batches, err = page_openai(openai_key, args.max_pages)
        if err:
            log.warning("openai batch list stopped early: %s", err)
        rows.extend(openai_cancel_rows(batches))
    if anthropic_key:
        checked.append("anthropic")
        batches, err = page_anthropic(anthropic_key, args.max_pages)
        if err:
            log.warning("anthropic batch list stopped early: %s", err)
        rows.extend(anthropic_cancel_rows(batches))

    stuck = stuck_rows(rows, now, max(1, args.stuck_minutes) * 60)
    salvage = salvage_rows(rows)
    stuck_ids = {r["id"] for r in stuck}

    for r in rows:
        log.info("%-11s %-14s %-11s %s of %s done, %s stopped", r["provider"],
                 r["id"][:14], r["status"], format(r["done"], ","),
                 format(r["total"], ","), format(r["stopped"], ","))
        if r["artifact"]:
            label = "output_file_id" if r["provider"] == "openai" else "results_url"
            log.info("%-11s %-14s   %s present", "", "", label)
        if r["id"] in stuck_ids:
            started = r.get("cancel_started")
            age = "an unknown time" if started is None \
                else "%d min" % ((now - started) // 60)
            log.warning("%-11s %-14s   mid cancel for %s", r["provider"],
                        r["id"][:14], age)

    state, detail = verdict(rows, stuck, salvage)
    emit = log.warning if state in FINDINGS else log.info
    emit("%-20s %s", state, detail)
    emit("  checked: %s", ", ".join(checked))
    emit("  measured: request_counts and the cancellation timestamps from the "
         "batch lists")
    emit("  inferred: that a re-run would repeat the finished rows, since "
         "neither API records whether the partial output was downloaded")
    for line in repair_lines(state, rows):
        emit("  repair: %s", line)

    total = len(stuck) + len(salvage)
    log.info("%d finding(s)", total)
    return 1 if total else 0


if __name__ == "__main__":
    sys.exit(main())

File: batch-cancelled-partial-results/node/batch-cancellation-audit.mjs
/**
 * Find billed, salvageable output left behind by cancelled batches.
 *
 * Read only, on both providers: /v1/batches on OpenAI and /v1/messages/batches
 * on Anthropic. Nothing is cancelled, submitted or downloaded.
 *
 * Cancel is a stop, not a rollback. Anthropic documents that canceled and
 * expired requests are not billed; OpenAI documents the partial output but not
 * the billing split, so its completed count is reported as a floor.
 */
const OPENAI_BATCHES_URL = 'https://api.openai.com/v1/batches';
const ANTHROPIC_BATCHES_URL = 'https://api.anthropic.com/v1/messages/batches';

export const STUCK_SECONDS = 15 * 60;

const OPENAI_CANCEL_STATES = new Set(['cancelling', 'cancelled']);

const FINDINGS = new Set(['cancel-stuck', 'cancel-partial-unclaimed']);

/** Epoch seconds from a unix number or an RFC 3339 string. Pure. */
export function parseTime(value) {
  if (value === null || value === undefined || value === '' || typeof value === 'boolean') {
    return null;
  }
  if (typeof value === 'number') return Math.trunc(value);
  const ms = Date.parse(String(value));
  return Number.isFinite(ms) ? Math.floor(ms / 1000) : null;
}

/** Normalised rows for OpenAI batches under cancellation. Pure. */
export function openaiCancelRows(batches) {
  return (batches ?? [])
    .filter((b) => OPENAI_CANCEL_STATES.has((b ?? {}).status))
    .map((b) => {
      const counts = b.request_counts ?? {};
      const total = Number(counts.total) || 0;
      const done = Number(counts.completed) || 0;
      const failed = Number(counts.failed) || 0;
      return {
        provider: 'openai',
        id: String(b.id),
        status: b.status,
        inFlight: b.status === 'cancelling',
        done,
        stopped: Math.max(0, total - done - failed),
        total,
        artifact: b.output_file_id ?? null,
        cancelStarted: parseTime(b.cancelling_at),
        billingKnown: false,
      };
    })
    .sort((a, b) => a.id.localeCompare(b.id));
}

/** Normalised rows for Claude batches under cancellation. Pure. */
export function anthropicCancelRows(batches) {
  return (batches ?? [])
    .filter((b) => (b ?? {}).cancel_initiated_at)
    .map((b) => {
      const counts = b.request_counts ?? {};
      const done = Number(counts.succeeded) || 0;
      const stopped = Number(counts.canceled) || 0;
      const status = String(b.processing_status ?? '') || 'unknown';
      return {
        provider: 'anthropic',
        id: String(b.id),
        status,
        inFlight: status === 'canceling',
        done,
        stopped,
        total: done + stopped + (Number(counts.errored) || 0)
               + (Number(counts.expired) || 0) + (Number(counts.processing) || 0),
        artifact: b.results_url ?? null,
        cancelStarted: parseTime(b.cancel_initiated_at),
        billingKnown: true,
      };
    })
    .sort((a, b) => a.id.localeCompare(b.id));
}

/** Rows holding finished work a re-run would pay for again. Pure. */
export function salvageRows(rows) {
  return (rows ?? []).filter((r) => (Number(r?.done) || 0) > 0);
}

/** Rows still mid cancel past the threshold. Pure. now is an argument. */
export function stuckRows(rows, now, seconds = STUCK_SECONDS) {
  return (rows ?? []).filter((r) => {
    if (!r?.inFlight) return false;
    const started = r.cancelStarted;
    return started === null || started === undefined || now - started > seconds;
  });
}

/** Finished rows across everything cancelled. Pure. */
export function salvagedTotal(rows) {
  return salvageRows(rows).reduce((n, r) => n + (Number(r.done) || 0), 0);
}

/** Grade the run. Pure. Returns [state, detail]. */
export function verdict(rows, stuck, salvage) {
  const all = rows ?? [];
  const s = stuck ?? [];
  const v = salvage ?? [];
  if (!all.length) {
    return ['no-cancels',
      'no batch on the providers checked has had a cancellation initiated'];
  }
  if (s.length) {
    let detail = `${s.length} batch(es) have been mid cancel longer than the `
      + 'documented window';
    if (v.length) {
      detail += `, and ${v.length} cancelled batch(es) hold ${salvagedTotal(v)} `
        + 'finished rows nothing has collected';
    }
    return ['cancel-stuck', detail];
  }
  if (v.length) {
    return ['cancel-partial-unclaimed',
      `${v.length} cancelled batch(es) hold ${salvagedTotal(v)} finished rows a `
      + 're-run would pay for again'];
  }
  return ['cancel-clean',
    `${all.length} cancellation(s) found, none of which had completed a single `
    + 'request, so there is nothing to salvage and nothing to double pay'];
}

/** The repair for one verdict. Pure. Printed, never performed. */
export function repairLines(state, rows) {
  const all = rows ?? [];
  if (state === 'no-cancels') return [];
  if (state === 'cancel-clean') {
    return ['nothing to collect. Keep cancelling early: a batch stopped before '
      + 'its first request completed costs nothing.'];
  }
  const lines = [];
  if (state === 'cancel-stuck') {
    lines.push('a batch still in cancelling or canceling has not stopped. Poll '
      + 'it to a terminal state before you submit a replacement, or the two '
      + 'will run the same rows at once.');
  }
  if (all.some((r) => (Number(r?.done) || 0) > 0)) {
    lines.push('download the partial output, drop those custom_ids from the '
      + 'input file, and submit only the remainder. Results are not returned in '
      + 'request order, so custom_id is the only join key that works.');
  }
  if (all.some((r) => r?.provider === 'anthropic' && (Number(r.done) || 0) > 0)) {
    lines.push('on Anthropic, canceled and expired requests are not billed, so '
      + 'the succeeded count is the whole cost of the cancelled batch.');
  }
  if (all.some((r) => r?.provider === 'openai' && (Number(r.done) || 0) > 0)) {
    lines.push('on OpenAI the billing split for a cancelled batch is not '
      + 'documented, so treat the completed count as a floor and confirm the '
      + 'day against the cost report.');
  }
  return lines;
}

async function getJson(url, headers, params) {
  const target = new URL(url);
  for (const [k, v] of Object.entries(params ?? {})) target.searchParams.set(k, String(v));
  let res;
  try {
    res = await fetch(target, { headers });
  } catch (err) {
    return [null, `request failed: ${err.message}`];
  }
  if (res.status !== 200) return [null, `HTTP ${res.status}`];
  try {
    return [await res.json(), null];
  } catch {
    return [null, 'response was not JSON'];
  }
}

async function pageOpenai(key, maxPages) {
  const rows = [];
  let after = null;
  for (let i = 0; i < Math.max(1, maxPages); i += 1) {
    const params = { limit: 100 };
    if (after) params.after = after;
    const [payload, err] = await getJson(OPENAI_BATCHES_URL,
      { Authorization: `Bearer ${key}` }, params);
    if (err) return [rows, err];
    const data = payload.data ?? [];
    rows.push(...data);
    if (!payload.has_more || !data.length) break;
    after = data[data.length - 1]?.id;
  }
  return [rows, null];
}

async function pageAnthropic(key, maxPages) {
  const rows = [];
  let after = null;
  for (let i = 0; i < Math.max(1, maxPages); i += 1) {
    const params = { limit: 1000 };
    if (after) params.after_id = after;
    const [payload, err] = await getJson(ANTHROPIC_BATCHES_URL,
      { 'x-api-key': key, 'anthropic-version': '2023-06-01' }, params);
    if (err) return [rows, err];
    const data = payload.data ?? [];
    rows.push(...data);
    if (!payload.has_more || !data.length) break;
    after = payload.last_id ?? data[data.length - 1]?.id;
  }
  return [rows, null];
}

function args(argv) {
  const out = { stuckMinutes: 15, maxPages: 20 };
  for (let i = 0; i < argv.length; i += 1) {
    if (argv[i] === '--stuck-minutes') out.stuckMinutes = Number.parseInt(argv[i += 1], 10);
    else if (argv[i] === '--max-pages') out.maxPages = Number.parseInt(argv[i += 1], 10);
  }
  return out;
}

async function main() {
  const opts = args(process.argv.slice(2));
  const openaiKey = (process.env.OPENAI_API_KEY || "dummy-openai-api-key");
  const anthropicKey = (process.env.ANTHROPIC_API_KEY || "dummy-anthropic-api-key");
  if (!openaiKey && !anthropicKey) {
    console.error('set OPENAI_API_KEY (project key, Read Only) or '
      + 'ANTHROPIC_API_KEY (workspace key), or both');
    process.exitCode = 2;
    return;
  }

  const now = Math.floor(Date.now() / 1000);
  const rows = [];
  const checked = [];
  if (openaiKey) {
    checked.push('openai');
    const [batches, err] = await pageOpenai(openaiKey, opts.maxPages);
    if (err) console.log(`openai batch list stopped early: ${err}`);
    rows.push(...openaiCancelRows(batches));
  }
  if (anthropicKey) {
    checked.push('anthropic');
    const [batches, err] = await pageAnthropic(anthropicKey, opts.maxPages);
    if (err) console.log(`anthropic batch list stopped early: ${err}`);
    rows.push(...anthropicCancelRows(batches));
  }

  const stuck = stuckRows(rows, now, Math.max(1, opts.stuckMinutes) * 60);
  const salvage = salvageRows(rows);
  const stuckIds = new Set(stuck.map((r) => r.id));

  for (const r of rows) {
    console.log(`${r.provider.padEnd(11)} ${r.id.slice(0, 14).padEnd(14)} `
      + `${r.status.padEnd(11)} ${r.done} of ${r.total} done, ${r.stopped} stopped`);
    if (r.artifact) {
      const label = r.provider === 'openai' ? 'output_file_id' : 'results_url';
      console.log(`${''.padEnd(27)} ${label} present`);
    }
    if (stuckIds.has(r.id)) {
      const age = r.cancelStarted === null || r.cancelStarted === undefined
        ? 'an unknown time'
        : `${Math.floor((now - r.cancelStarted) / 60)} min`;
      console.log(`${r.provider.padEnd(11)} ${r.id.slice(0, 14).padEnd(14)}   `
        + `mid cancel for ${age}`);
    }
  }

  const [state, detail] = verdict(rows, stuck, salvage);
  console.log(`${state.padEnd(20)} ${detail}`);
  console.log(`  checked: ${checked.join(', ')}`);
  console.log('  measured: request_counts and the cancellation timestamps from '
    + 'the batch lists');
  console.log('  inferred: that a re-run would repeat the finished rows, since '
    + 'neither API records whether the partial output was downloaded');
  for (const line of repairLines(state, rows)) console.log(`  repair: ${line}`);

  const total = stuck.length + salvage.length;
  console.log(`${total} finding(s)`);
  process.exitCode = total ? 1 : 0;
  void FINDINGS;
}

if (import.meta.url === `file://${process.argv[1]}`) await main();
